### PR TITLE
Evo2 merge 20250214

### PR DESCRIPTION
**Pull request by JRD971000**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/tokenizers/bytelevel_tokenizers.py [hunk 1]
```diff
@@ -26,86 +26,73 @@ class ByteLevelProcessor:
...
-    def bos_id(self):
-        return 257
-
-    @property
-    def eos_id(self):
-        return 258
-
-    @property
-    def unk_id(self):
-        return 259  # unused
```
Comment (cspades) [inline nemo/collections/common/tokenizers/bytelevel_tokenizers.py:111]: Shouldn't be deleted, seems like some rebase or checkout error?

File: nemo/collections/common/tokenizers/bytelevel_tokenizers.py [hunk 2]
```diff
@@ -26,86 +26,73 @@ class ByteLevelProcessor:
...
+                Defaults to 512.
+            _eos_id: ID to use for the end-of-sequence token.
+                Defaults to 0.
+            _pad_id: ID to use for the padding token.
+                Defaults to 1.
+            _bos_id: ID to use for the beginning-of-sequence token.
+                Defaults to None.
+        """
+        self.vocab_size = vocab_size if special_tokens is None else vocab_size + len(special_tokens)
+        self.special_start = vocab_size
```
Comment (cspades) [inline nemo/collections/common/tokenizers/bytelevel_tokenizers.py:82]: This is a duplicate of L91-93, I changed the order of when this was called when I pulled these changes in.

File: nemo/collections/common/tokenizers/bytelevel_tokenizers.py (hunk 1 above)
Comment (jstjohn) [reply nemo/collections/common/tokenizers/bytelevel_tokenizers.py:111]: Good find.

File: nemo/collections/llm/gpt/model/hyena.py [hunk 3]
```diff
@@ -0,0 +1,698 @@
...
+            HyenaConfig: Model configuration
+        """
+        return self.model_config
+
+
+@io.state_transform(
+    source_key=("sequential.*.mlp.w1.weight", "sequential.*.mlp.w2.weight"),
+    target_key="decoder.layers.*.mlp.linear_fc1.weight",
+)
+def _import_linear_fc1(w1, w2):
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/model/hyena.py:510]: can you switch to using TransformFns.merge_fc1?

File: nemo/collections/llm/gpt/model/hyena.py [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+@dataclass
+class Hyena7bARCLongContextConfig(Hyena7bConfig):
+    """The checkpoint from ARC requires padding to the FFN dim
+    due to constraintes from large TP size for training."""
+
+    ffn_hidden_size: int = 11264
+
+
+@dataclass
+class Hyena40bARCLongContextConfig(Hyena40bConfig):
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/model/hyena.py:668]: Up to you but should we move these config classes up before the importer?

File: nemo/lightning/io/registry.py [hunk 5]
```diff
@@ -48,11 +48,20 @@
...
     track_io(TiktokenTokenizer, artifacts=[FileArtifact("vocab_file")])
+    track_io(ByteLevelTokenizer)
+except ImportError:
+    # Tokenizers are not available, no need to track it.
+    pass
+
+try:
+    from lightning.pytorch.loggers import TensorBoardLogger, WandbLogger
+
+    track_io(WandbLogger)
```
Comment (hemildesai) [inline nemo/lightning/io/registry.py:64]: Any reason why this is needed? I think we have used both Wandb and Tensorboard without tracking io before.

File: nemo/lightning/pytorch/callbacks/flops_callback.py [hunk 6]
```diff
@@ -32,6 +34,7 @@
     "nemotron": flops_formulas.nemotron,
     "mixtral": flops_formulas.mixtral,
     "bert": flops_formulas.bert,
+    "hyena": hyena,
```
Comment (hemildesai) [inline nemo/lightning/pytorch/callbacks/flops_callback.py:37]: Can we just add this to flop formulas and keep it consistent with the above?

Comment (jstjohn) [reply nemo/lightning/pytorch/callbacks/flops_callback.py:37]: My goal here was to keep the arc+nvidia copyright header out of our shared flops formula file.

File: nemo/lightning/io/registry.py (hunk 5 above)
Comment (jstjohn) [reply nemo/lightning/io/registry.py:64]: @cspades or @JRD971000 do either of you remember why we added this? Maybe it's not needed.

File: nemo/collections/llm/gpt/model/hyena.py (hunk 4 above)
Comment (jstjohn) [reply nemo/collections/llm/gpt/model/hyena.py:790]: Good idea, I'll do that next week.

File: nemo/collections/llm/gpt/model/hyena.py (hunk 3 above)
Comment (jstjohn) [reply nemo/collections/llm/gpt/model/hyena.py:587]: Good idea. Thanks!

File: nemo/collections/llm/gpt/data/megatron/__init__.py
[the remark this answers is not in the document]
Comment (jstjohn) [reply nemo/collections/llm/gpt/data/megatron/__init__.py:1]: These are sub-classes of megatron data modules that were decided shouldn't go into the megatron-lm repo.

File: tests/collections/llm/gpt/model/test_hyena.py [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+        num_nodes=args.num_nodes,
+        max_steps=args.max_steps,
+        accelerator="gpu",
+        strategy=strategy,
+        logger=loggers,
+        callbacks=callbacks,
+        log_every_n_steps=args.log_every_n_steps,
+        limit_val_batches=args.limit_val_batches,
+        num_sanity_val_steps=0,
+        use_distributed_sampler=True,
```
Comment (jstjohn) [inline tests/collections/llm/gpt/model/test_hyena.py:596]: @JRD971000 are you sure this should be True? I thought that this was to use the lightning distributed sampler, but we use a custom megatron sampler as part of our megatron strategy and data loaders. Did something change recently that requires us to set this to True?

File: nemo/lightning/io/registry.py (hunk 5 above)
Comment (cspades) [reply nemo/lightning/io/registry.py:64]: When I tried to run inference on a checkpoint produced by a training script using WandBLogger, the checkpoint load failed until I added this class to the IO!

What's should I do?

File: nemo/lightning/pytorch/callbacks/flops_callback.py [hunk 8]
```diff
@@ -43,7 +46,7 @@ class FLOPsMeasurementCallback(Callback):
         model_config (GPTConfig): Model parameters.
         data_config (pl.LightningDataModule): Data module being used in the experiment.
         model_name (str): Name of the model being run. The following models are supported:
-            gpt3, llama2, llama3, nemotron, mixtral, bert.
+            gpt3, llama2, llama3, nemotron, mixtral, bert, hyena.
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply nemo/lightning/pytorch/callbacks/flops_callback.py:49]: Fixed, thanks @dorotat-nv

File: nemo/utils/hyena_flops_formulas.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            elif layer == "D":
+                D += 1
+            elif layer == "H":
+                H += 1
+            elif layer == "*":
+                A += 1
+        return S, D, H, A
+
+    # Count S, D, H, and * layers in HyenaModel.
+    S, D, H, A = _hyena_layer_count(config.model_pattern)
```
[the remark this answers is not in the document]
Comment (JRD971000) [reply nemo/utils/hyena_flops_formulas.py:56]: Fixed, thanks @dorotat-nv

**Review by ko3n1g — CHANGES_REQUESTED** [review]
Hey @jstjohn , there's a final job called `Nemo_CICD_Test` - could you add your newly onboarded job into it's `.needs` list? This will help the CI to compute the overall CI result. Everything else looks good to me.

**Review by ko3n1g — APPROVED** [review]

**Review by dorotat-nv — APPROVED** [review]
Approving on the condition that there will be a follow up PR with a cleanup

**Comment by JRD971000** [comment]
> Approving on the condition that there will be a follow up PR with a cleanup

Thanks @dorotat-nv , we'll have a separate following PR with the cleaups you suggested.

**Review by ko3n1g — DISMISSED** [review]

**Review by cuichenx — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

**Review by pablo-garay — APPROVED** [review]
